src/Middleware.php

```diff
@@ -3,31 +3,127 @@
 namespace Inertia;
 
 use Closure;
-use Illuminate\Support\Facades\Response;
-use Symfony\Component\HttpFoundation\RedirectResponse as Redirect;
+use Illuminate\Http\Request;
+use Illuminate\Support\Collection;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
 
-class Middleware
+abstract class Middleware
 {
-    public function handle($request, Closure $next)
+    /**
+     * Determines the current Inertia asset version hash.
+     *
+     * @see https://inertiajs.com/asset-versioning
+     * @param  \Illuminate\Http\Request  $request
+     * @return string|null
+     */
+    public function version(Request $request)
     {
+        //
+    }
+
+    /**
+     * Defines the Inertia props that are shared by default.
+     *
+     * @see https://inertiajs.com/shared-data
+     * @param  \Illuminate\Http\Request  $request
+     * @return array
+     */
+    public function share(Request $request)
+    {
+        return [
+            'errors' => function () use ($request) {
+                return $this->resolveValidationErrors($request);
+            },
+        ];
+    }
+
+    /**
+     * Handle the incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  Closure  $next
+     * @return Response
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        Inertia::share($this->share($request));
+
+        Inertia::version(function () use ($request) {
```

Comment:
Does this need to be in a closure, since this now only runs when the middleware runs?

Reply:
It doesn't, but it allows the user to not have to care about it in their `version()` function, and still have it lazy evaluate out of the box 🎉

Reply:
Right, but even for end users, if this is done in a middleware, why would they need it to be lazily evaluated if it's going to be immediately resolved?

Reply:
It won't, because the controller action still has to happen first, meaning that there might still be a case where a redirect happens instead of an Inertia response being rendered. 

In that scenario, resolving it immediately would mean unnecessarily loading a file & hashing it.

Reply:
Yup, okay, fair enough! 👍